### PR TITLE
Relax priv-in-pub lint on generic bounds and where clauses of trait impls.

### PR DESCRIPTION
**Pull request by jswrenn**
The priv-in-pub lint is a legacy mechanism of the compiler, supplanted by a reachability-based [type privacy](https://github.com/rust-lang/rfcs/blob/master/text/2145-type-privacy.md) analysis. This PR does **not** relax type privacy; it only relaxes the lint (as proposed by the type privacy RFC) in the case of trait impls.

## Current Behavior
On public trait impls, it's currently an **error** to have a `where` bound constraining a private type with a trait:
```rust
pub trait Trait {}
pub struct Type {}

struct Priv {}
impl Trait for Priv {}

impl Trait for Type
where
    Priv: Trait // ERROR
{}
```

...and it's a **warning** to have have a public type constrained by a private trait:
```rust
pub trait Trait {}
pub struct Type {}

pub struct Pub {}
trait Priv {}
impl Priv for Pub {}

impl Trait for Type
where
    Pub: Priv // WARNING
{}
```

This lint applies to `where` clauses in other contexts, too; e.g. on free functions:
```rust
struct Priv<T>(T);
pub trait Pub {}
impl<T: Pub> Pub for Priv<T> {}

pub fn function<T>()
where
    Priv<T>: Pub // WARNING
{}
```

**These constraints could be relaxed without issue.**

## New Behavior
This lint is relaxed for `where` clauses on trait impls, such that it's okay to have a `where` bound constraining a private type with a trait:
```rust
pub trait Trait {}
pub struct Type {}

struct Priv {}
impl Trait for Priv {}

impl Trait for Type
where
    Priv: Trait // OK
{}
```

...and it's okay to have a public type constrained by a private trait:
```rust
pub trait Trait {}
pub struct Type {}

pub struct Pub {}
trait Priv {}
impl Priv for Pub {}

impl Trait for Type
where
    Pub: Priv // OK
{}
```

## Rationale
While the priv-in-pub lint is not essential for soundness, it *can* help programmers avoid pitfalls that would make their libraries difficult to use by others. For instance, such a lint *is* useful for free functions; e.g. if a downstream crate tries to call the `function` in the previous snippet in a generic context:
```rust
fn callsite<T>()
where
    Priv<T>: Pub // ERROR: omitting this bound is a compile error, but including it is too
{
    function::<T>()
}
```
...it cannot do so without repeating `function`'s `where` bound, which we cannot do because `Priv` is out-of-scope. A lint for this case is arguably helpful.

However, this same reasoning **doesn't** hold for trait impls. To call an unconstrained method on a public trait impl with private bounds, you don't need to forward those private bounds, you can forward the public trait:
```rust
mod upstream {
    pub trait Trait {
        fn method(&self) {}
    }
    pub struct Type<T>(T);
    
    pub struct Pub<T>(T);
    trait Priv {}
    impl<T: Priv> Priv for Pub<T> {}
    
    impl<T> Trait for Type<T>
    where
        Pub<T>: Priv // WARNING
    {}
}

mod downstream {
    use super::upstream::*;
    
    fn function<T>(value: Type<T>)
    where
        Type<T>: Trait // <- no private deets!
    {
        value.method();
    }
}
```

**This PR only eliminates the lint on trait impls.** It leaves it intact for all other contexts, including trait definitions, inherent impls, and function definitions. It doesn't need to exist in those cases either, but I figured I'd first target a case where it's mostly pointless.

## Other Notes
- See discussion [on zulip](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/relax.20priv-in-pub.20lint.20for.20trait.20impl.20.60where.60.20bounds/near/222458397).
- This PR effectively reverts #79291.

### CODE REVIEW

**Comment by petrochenkov** [comment]
I'm going to send this to the lang team for sanity checking.

This PR removes the private-in-public warnings from generic bounds and where clauses on `impl` items specifically.
- The behavior introduced by this PR matches the "Type privacy" RFC precisely if the `private_bounds` lint (https://github.com/rust-lang/rfcs/blob/master/text/2145-type-privacy.md#lint-2-private-traitstypes-in-secondary-interface-of-effectively-public-items) ends up being allow-by-default (which seems somewhat reasonable), or will similarly not be reported for `impl` items specifically.
- The behavior introduced by this PR will differ from the "Type privacy" RFC, if the `private_bounds` lint ends up being warn-by-default, in that case code allowed by this PR will start reporting warnings.

**Comment by nikomatsakis** [comment]
@petrochenkov I cannot quite tell from your comment what your take is here -- do you feel this is in the spirit of the original RFC? In what way does it differ? It seems surprising to me that you would say it conforms or not to the RFC based on whether the lint in the RFC is allow by default.

**Comment by petrochenkov** [comment]
>do you feel this is in the spirit of the original RFC?

It's in the spirit of the RFC.
The RFC is just not entirely precise

>This lint replaces part of private-in-public errors and can be reported [**as warn-by-default or allow-by-default**](https://github.com/rust-lang/rfcs/blob/master/text/2145-type-privacy.md#lint-2-private-traitstypes-in-secondary-interface-of-effectively-public-items).

Back then I wanted to start with warn-by-default to get some experience, but maybe be we already have enough experience to just allow it now.

**Comment by jswrenn** [comment]
This PR also draws a finer distinction than [the proposed Lint #2 makes](https://github.com/rust-lang/rfcs/blob/master/text/2145-type-privacy.md#lint-2-private-traitstypes-in-secondary-interface-of-effectively-public-items): this PR distinguishes between private items in the bounds of trait impls, and private items in other bounds. I describe the rationale for this distinction in the description of this PR, and this PR only impacts private items in the `where` bounds of trait impls.

---

@scottmcm [suggested](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/relax.20priv-in-pub.20lint.20for.20trait.20impl.20.60where.60.20bounds/near/222483732) that all uses of private items in bounds should be allowed, but this PR doesn't go that far. I'm inclined to agree, though. The only danger of private items in bounds is that end-users can't 'forward' such bounds when they're writing generic code. But:
- not all bounds *need* to be forwarded
- you can already write un-forwardable bounds, just by adding a tad of indirection
- sometimes unforwardable bounds are desirable

So, the lint proposed proposed by the type privacy RFC would just a heuristic that _sometimes_ helps avoid an API design decision that _sometimes_ is a pitfall. There's enough "sometimes"s in there, that I'd argue such a lint would probably be more at home in clippy than rustc.

**Comment by scottmcm** [comment]
FWIW, I like the middle-ground proposed here.  The OP did a good job convincing me that the "you can forward the public trait" is an interesting distinction.  We can always remove more cases later.

**Comment by scottmcm** [comment]
We didn't have time to go into this in detail in today's lang team meeting, so let's try to make progress async:

@rfcbot fcp merge

The OP here has a good description of what this allows.  People seem to want to do this enough that the workarounds are well-known; I think we should just allow it directly.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

File: src/test/ui/privacy/where-priv-type.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
+#![crate_type = "lib"]
```
Comment (petrochenkov) [inline src/test/ui/privacy/where-priv-type.rs:1]: Could you add a comment explaining what this test is testing in a couple of sentences?

Comment (petrochenkov) [reply src/test/ui/privacy/where-priv-type.rs:1]: (And to the second new test too.)

File: src/test/ui/privacy/private-in-public-warn.rs [hunk 2]
```diff
@@ -85,11 +83,7 @@ mod traits_where {
         //~| WARNING hard error
     }
     impl<T> Pub<T> where T: PrivTr {}
-        //~^ ERROR private trait `traits_where::PrivTr` in public interface
-        //~| WARNING hard error
     impl<T> PubTr for Pub<T> where T: PrivTr {}
```
Comment (petrochenkov) [inline src/test/ui/privacy/private-in-public-warn.rs:86]:
```suggestion
    impl<T> PubTr for Pub<T> where T: PrivTr {} // OK, trait impl predicates
```

File: src/test/ui/privacy/private-in-public-warn.rs [hunk 3]
```diff
@@ -61,10 +61,8 @@ mod traits {
         fn f<T: PrivTr>(arg: T) {} //~ ERROR private trait `traits::PrivTr` in public interface
         //~^ WARNING hard error
     }
-    impl<T: PrivTr> Pub<T> {} //~ ERROR private trait `traits::PrivTr` in public interface
-        //~^ WARNING hard error
-    impl<T: PrivTr> PubTr for Pub<T> {} //~ ERROR private trait `traits::PrivTr` in public interface
-        //~^ WARNING hard error
+    impl<T: PrivTr> Pub<T> {}
+    impl<T: PrivTr> PubTr for Pub<T> {}
```
Comment (petrochenkov) [inline src/test/ui/privacy/private-in-public-warn.rs:65]:
```suggestion
    impl<T: PrivTr> PubTr for Pub<T> {} // OK, trait impl predicates
```

File: src/test/ui/const-generics/generic_const_exprs/eval-privacy.rs [hunk 4]
```diff
@@ -10,11 +10,6 @@ pub trait Trait {
 }
 
 impl<const U: u8> Trait for Const<U>
```
Comment (petrochenkov) [inline src/test/ui/const-generics/generic_const_exprs/eval-privacy.rs:12]:
```suggestion
impl<const U: u8> Trait for Const<U> // OK, trait impl predicates
```

File: compiler/rustc_privacy/src/lib.rs [hunk 5]
```diff
@@ -2069,7 +2069,6 @@ impl<'tcx> Visitor<'tcx> for PrivateItemsInPublicInterfacesVisitor<'tcx> {
             // Subitems of trait impls have inherited publicity.
             hir::ItemKind::Impl(ref impl_) => {
                 let impl_vis = ty::Visibility::of_impl(item.def_id, tcx, &Default::default());
-                self.check(item.def_id, impl_vis).generics().predicates();
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:2072]: This change affects inherent impls too, which contradicts the PR description.

Comment (petrochenkov) [reply compiler/rustc_privacy/src/lib.rs:2072]: Also, could you add a comment saying that generic parameters and where clauses are specifically NOT checked, and it's not an accidental omission, and link to this PR.

Comment (jswrenn) [reply compiler/rustc_privacy/src/lib.rs:2072]: Fixed with https://github.com/rust-lang/rust/pull/90586/commits/6a5c0d6039e6095edfb302b4902997b76c46ce35!

File: src/test/ui/privacy/where-priv-type.rs (hunk 1 above)
Comment (jswrenn) [reply src/test/ui/privacy/where-priv-type.rs:4]: Resolved by https://github.com/rust-lang/rust/pull/90586/commits/932a34ce2005ccfae225aab2bf424e7233d54e48.

File: src/test/ui/privacy/private-in-public-warn.rs [hunk 6]
```diff
@@ -85,11 +84,9 @@ mod traits_where {
         //~| WARNING hard error
     }
     impl<T> Pub<T> where T: PrivTr {}
-        //~^ ERROR private trait `traits_where::PrivTr` in public interface
-        //~| WARNING hard error
-    impl<T> PubTr for Pub<T> where T: PrivTr {}
-        //~^ ERROR private trait `traits_where::PrivTr` in public interface
+        //~^ ERROR private trait `traits_where::PrivTr` in public interfac
```
Comment (petrochenkov) [inline src/test/ui/privacy/private-in-public-warn.rs:87]:
```suggestion
        //~^ ERROR private trait `traits_where::PrivTr` in public interface
```

File: src/test/ui/privacy/private-in-public.rs [hunk 7]
```diff
@@ -31,7 +31,7 @@ mod traits {
     pub enum E<T: PrivTr> { V(T) } //~ ERROR private trait `traits::PrivTr` in public interface
     pub fn f<T: PrivTr>(arg: T) {} //~ ERROR private trait `traits::PrivTr` in public interface
     pub struct S1<T: PrivTr>(T); //~ ERROR private trait `traits::PrivTr` in public interface
-    impl<T: PrivTr> Pub<T> { //~ ERROR private trait `traits::PrivTr` in public interface
+    impl<T: PrivTr> Pub<T> {  //~ ERROR private trait `traits::PrivTr` in public interface
```
Comment (petrochenkov) [inline src/test/ui/privacy/private-in-public.rs:34]: Nit: unnecessary whitespace change.

**Comment by petrochenkov** [comment]
r=me after squashing commits and removing accidental changes from the resulting commit.

**Review by petrochenkov — APPROVED** [review]

**Comment by ehuss** [comment]
@bors retry

network error
